packages/sync/src/webrtc-http-stream-signaling.ts

```diff
@@ -24,148 +24,161 @@ import * as buffer from 'lib0/buffer';
-									const existingConn = webrtcConns.get(
-										data.from
-									);
-									if ( existingConn ) {
-										existingConn.glareToken = undefined;
+					const emitPeerChange = webrtcConns.has( data.from )
+						? () => {}
+						: () =>
+								room.provider.emit( 'peers', [
+									{
+										removed: [],
+										added: [ data.from ],
+										webrtcPeers: Array.from(
+											room.webrtcConns.keys()
+										),
+										bcPeers: Array.from( room.bcConns ),
+									},
+								] );
+					switch ( data.type ) {
+						case 'announce':
+							if ( webrtcConns.size < room.provider.maxConns ) {
+								map.setIfUndefined(
+									webrtcConns,
+									data.from,
+									() =>
+										new WebrtcConn(
+											signalCon as SignalingConn,
+											true,
+											data.from,
+											room
+										)
+								);
+								emitPeerChange();
+							}
+							break;
+						case 'signal':
+							if ( data.signal.type === 'offer' ) {
+								const existingConn = webrtcConns.get(
+									data.from
+								);
+								if ( existingConn ) {
+									const remoteToken = data.token;
+									const localToken = existingConn.glareToken;
+									if (
+										localToken &&
+										localToken > remoteToken
+									) {
+										log( 'offer rejected: ', data.from );
+										return;
 									}
+									// if we don't reject the offer, we will be accepting it and answering it
+									existingConn.glareToken = undefined;
 								}
-								if ( data.to === peerId ) {
-									map.setIfUndefined(
-										webrtcConns,
-										data.from,
-										() =>
-											new WebrtcConn(
-												signalCon,
-												false,
-												data.from,
-												room
-											)
-									).peer.signal( data.signal );
-									emitPeerChange();
+							}
+							if ( data.signal.type === 'answer' ) {
+								log( 'offer answered by: ', data.from );
+								const existingConn = webrtcConns.get(
+									data.from
+								);
+								if ( existingConn ) {
+									existingConn.glareToken = undefined;
 								}
-								break;
-						}
-					};
-					if ( room.key ) {
-						if ( typeof m.data === 'string' ) {
-							cryptoutils
-								.decryptJson(
-									buffer.fromBase64( m.data ),
-									room.key
-								)
-								.then( execMessage );
-						}
-					} else {
-						execMessage( m.data );
+							}
+							if ( data.to === peerId ) {
+								map.setIfUndefined(
+									webrtcConns,
+									data.from,
+									() =>
+										new WebrtcConn(
+											signalCon as SignalingConn,
+											false,
+											data.from,
+											room
+										)
+								).peer.signal( data.signal );
+								emitPeerChange();
+							}
+							break;
 					}
+				};
+				if ( room.key ) {
+					if ( typeof m.data === 'string' ) {
+						cryptoutils
+							.decryptJson(
+								buffer.fromBase64( m.data ),
+								room.key
+							)
+							// @ts-ignore
```

Comment:
Can we please instead use `// @ts-expect-error` with a comment explaining why?

Also, if it's about the argument, then I would prefer moving the comment inside the parentheses to apply it appropriately.

```ts
// @ts-expect-error
.then( execMessage );
```
vs

```ts
.then(
    // @ts-expect-error
    execMessage
);
```

Former ignores a wide ranges of TS errors. For example, if `then` is not a function, it won't error, which is what we probably don't want.